### DISCUSSION TITLE
Public Keys in asset description JSON

### DISCUSSION THREAD

**Discussion by B0BSmiths** [CounterpartyXCP/CIPs#89] [General]
I would like to suggest adding support for additional public keys in the enhanced asset info JSON file - there is a proposed update in CIP 25.  PGP is mentioned in the v1 spec

BIP47 Paynyms and Nostr public keys could both be useful additions to the suggested upgrade and to the enhanced asset information specification.

BIP47 paynyms are now available in Sparrow desktop wallet as well as in Samourai mobile wallet.  BIP47 payment codes provide unique identities that allow pairs of wallets to create addresses known only to each other, which allows for more private use of bitcoin and could see counterparty assets serv as a namespace (via asset registrations, much like .eth can) for sending BTC privately without needing to request a XPUB or a new address from the recipient each time.

Creating assets.xcp as domains has been demonstrated but if these "identities" are used for sending bitcoin that will result in address reuse, having a BIP47 payment code in the enhanced asset information JSON creates a path to facilitate private payments between two parties whilst using the .xcp namespace

Nostr npubs are another public key that could be useful in the enhanced asset information.  This would make it easier for people to get in touch with the owner of an asset if a npub was present in the .json file

The suggested upgrade is proposing to allow linking to Facebook Twitter and other social media platforms but all of those are subject to arbitrary censorship, whereas Nostr is not and Nostr functions very much like bitcoin in that it has permission-less access and is accessed via private public key pairs

I expect other public keys could be useful additions in future so rather than just support for the two i have mentioned above we should add support for multiple user defined keys.  Perhaps someone would like to link an Ethereum public key to a counterparty asset ?

**Comment by jdogresorg** [comment]
Sound logic. I am supportive of a way to pass pub keys in a standardized format in JSON. We currently have `pgpsig` for passing signature data, which is very limited. How would you like to expand on that? 

Can you give an example of what you would like the JSON to look like?

```
"pubkeys":[{
	"type": "pgp",
	"data": "0C859EE8BC7D0A56055A6D312411DE071B0002BF"
},{
	...	
}]
```

What `type` of signatures could you think of that ppl would want? Should we have a `type` and `service` field? One to define type of key and one to define service key is related to?

**Reply by B0BSmiths** [reply]
I think people should be able to specify the various pubkeys they wish to use, I expect the available options grows with time and not everyone will want to use all available options.  I can also imagine people would like to link more than one of the same type of key - for example maybe on a token there is a linked company pubkey for nostr and then also a linked product specific pubkey also for nostr, so the format could be much as you have for the other enhancements

Another idea is to include hashes and signatures as optional parameters, signatures as those generated by freewallet Sign text message feature can protect the data and mitigates an evil maid attack on the .json file if it is hosted with a third party

"pubkeys": [{
        "type": "BIP47",
        "data": "PM8TJJgnizNyRWCasRK5HLVdpKVDzux8XrCta6CTnSsovScd2grRjJorqtcfUaLWKqn23xHtyGMBsotYCgUcHonV5gReqnwZrZqqZtDJVNTjsmrCW4eA",
        "sig": "IApOw/kyMAFnnMsDOIqk0RQT8Q8ov4bbSPZZMDTfeYf0N4GOfIiRRuQvU5VM9Kw471Z//Pujs3yWDzLVg8kQJ1M="
    },{
        "type": "nostr",
        "data": "npub1umh9k3yuygx7l6nrwwu20c28e27k0s4ak5qk3p4lvzt2836rtfsse626me"
        "sig": "IEQTJi+P5lKVW3orAY2HZiI4zydTV9iyIadI6zGBoo1JGLqv2tIdctciP6bmK+JjyJ2FLnqWD+KshncMLUyEZLU="
    },{
        "type": "nostr",
        "data": "npub1xwwhsp9k56dhauz6z6wh9j37jalkf6cq4dhwmus67z3vyvnkjxesf3m33n"
        "sig": "IEWuGHwkq+fVCzFN2ebsB7+xXNMEDvBXX0uISf8sEWIqfAdTryoQXDmHydwenkaUkUTbFITuw5H2o4FEhA5hFFw="
    },{
        "type": "nostr",
        "data": "npub1qny3tkh0acurzla8x3zy4nhrjz5zd8l9sy9jys09umwng00manysew95gx"
        "sig": "IEQTJi+P5lKVW3orAY2HZiI4zydTV9iyIadI6zGBoo1JGLqv2tIdctciP6bmK+JjyJ2FLnqWD+KshncMLUyEZLU="
    },
    {
        "type": "eth",
        "data": "0xb794f5ea0ba39494ce839613fffba74279579268"
        "sig": "H2gm8Sab9DiPxQ/TcaS23wx25ruFmTennPPvlrSFjrbvGKB8jiXnJYalC1q1cX6o++zWBQVss21vcfDo06EWS8c="
    },    {
        "type": "doge",
        "data": "D9TQ3Y9k3Pr8LneX9quS1Ppckw5kREEoZc"
        "sig": "H3eIz+Rdyj5cweO1hyssGqvG/RL1HP1IQVH43+0aX0WVImKcG++cALHA4641v4hTqV/v/9K7J8YH6ix0Pi/tcHc="
    }],

**Comment by jotapea** [comment]
I have issues with CIP25 😆. Mainly because it is not about the protocol at all. Users should be free to use whatever JSON format they want.

But I wouldn't mind having a place where the CIP25 format (which I would rename to "XChain2022" or something like that) is put side by side with the classic format and any other formats / standards actually used by JSON-in-url-in-description-in-asset (written like this to show how disconnected these are from the protocol).

The protocol should not be taking any position in how something external from the protocol is done.

**Comment by jdogresorg** [comment]
CIP25 is about extending the set of standard JSON fields that the founders established in the enhanced asset info spec, to allow block explorers to parse in the data in a standardized way. Users are, and always have been, able to use ANY JSON file format. However, without standardizing the fields to be used, there is no sane way to parse the data in and display it.

Thanks to CIP25 standardization of fields, we now are able to display audio, video, images, and much more in a standardized way.

Perhaps limiting your comments to just the specific topic at hand (should we add a new field to CIP25) would be helpful to keep the conversation focused and on point.

**Comment by jotapea** [comment]
@jdogresorg I don't think Counterparty data explorers need to follow ANY specific external JSON format. Counterparty data explorer should only require one thing: show the protocol data. Anything else external to the protocol should be up to what the user / explorer-provider wants.

Having an "official" central place to publish different kinds of JSON formats in use is different, and I can definitively agree with this. But not as a you-must-follow-this-standard JSON format. If there was going to be one, I would leave the original simple founders one only.

> Thanks to CIP25 standardization of fields, we now are able to display audio, video, images, and much more in a standardized way.

You mean XChain and that is cool. It should be renamed to the "XChain" format.

**Comment by Jpja** [comment]
@jotapea CIP25 is an **informational** proposal. CIP01 describes the purpose of this category:

_An Informational CIP describes a Counterparty design issue, or provides general guidelines or information to the Counterparty community, but does not propose a new feature. Informational CIPs do not necessarily represent a Counterparty community consensus or recommendation, so users and implementors are free to ignore Informational CIPs or follow their advice._

I.e. you are free to ignore it.

**Comment by jotapea** [comment]
@Jpja Thanks for that, but still, what would happen if another service provider proposes a new JSON format? Will this new provider also have the opportunity to add another informational proposal as a CIP?

The main point I want to bring is that the protocol repo should avoid taking a stand on how the protocol is USED. In my view, these external JSON links should not be part of the official protocol, thus CIPs.

So, can we at least agree that the "CIP25 format" should be renamed to the XChain format? I can agree to extending CIP25 to now be in the context of creating a dedicated space in the repo where service providers can submit their JSON standards. And XChain can keep the "spotlight" in the CIP as a first example. But the format itself is not named CIP25, it is named the "XChain format", the first example on CIP25 for a generalized XChain-page-like approach to an asset's (web) JSON.

Thus, no need for new JSON-link-in-description related CIPs.

If we can agree to this, I can work on this extension to CIP25.

**Reply by Jpja** [reply]
@jotapea I think it's best to refer to it as "CIP25 JSON" or "CIP25 version" or something unambiguously referring to CIP25. I don't see anything in this proposal that links to or favors xchain. The author only happens to run xchain. 

This _informational_ CIP for standardizing JSON files is very useful imo. Makes life easier for asset creators AND anyone building wallets/explorers.

**Reply by jotapea** [reply]
As someone building a web JSON aware [explorer](https://bitst.art), CIP25 actually made JSON logic more complex. All past JSONs still need to be parsed correctly, so @jdogresorg's "standard" is just one more way to do it.

The previous single level JSON standard (no arrays of objects) people had been using before is simpler, thus superior in my opinion and **should also have a place in the repo**. And even if some don't consider it superior I STILL believe it should have a place in the repo! Because it is widely used, in like almost all JSONs before CIP25 was introduced, and many after.

If anyone wants to make an XChain clone, then obviously "CIP25" will make their life "easier" (if you want to ONLY support the "CIP25" format). Thus why I believe it should be named the XChain format.

But is ok if my point is not getting across, is understandable how people still see XChain AS Counterparty. I don't.

I might do a separate discussion about all this some other time.
